server/util/cache_metrics/cache_metrics.go

```diff
@@ -0,0 +1,175 @@
+package cache_metrics
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+
+	gstatus "google.golang.org/grpc/status"
+)
+
+// CacheTier is an enumerated label value that describes the level at
+// which the cache sits in the composable cache stack.
+//
+// This label enables Prometheus queries that don't break when the cache implementation
+// is changed. For example, to get cache metrics for the in-memory tier only (without having
+// to explicitly refer to "redis", "memcached", etc.), one could write this query:
+//
+//     sum by (backend) (rate(buildbuddy_cache_write_count{tier="memory"}[5m])
+//
+// These types of queries are useful when monitoring the cache because caches
+// at different tiers will have observed latencies at different orders of magnitude,
+// so each tier likely needs its own chart in order to be observed at an appropriate
+// resolution.
+type CacheTier string
+
+const (
+	// MemoryCacheTier is used for in-memory cache implementations, like memcached and redis.
+	MemoryCacheTier CacheTier = "memory"
+	// CloudCacheTier is used for cloud storage cache implementations, like GCS or AWS S3.
+	CloudCacheTier CacheTier = "cloud"
+)
+
+func MakeCacheLabels(tier CacheTier, backend string) prometheus.Labels {
+	return prometheus.Labels{
+		metrics.CacheTierLabel:    string(tier),
+		metrics.CacheBackendLabel: backend,
+	}
+}
+
+func appendLabels(a prometheus.Labels, b prometheus.Labels) prometheus.Labels {
```

Comment:
Is the deduping even necessary? I peeked at the source for prom and wasn't sure -- didn't seem like it but you are the expert here.

If you don't need to dedupe, you can do away with this function entirely and in place of 
 appendLabels(a,b)
just use:
 append(a, b...)

Reply:
I think `prometheus.Labels` is a type alias for `map[string]string` so AFAIK I can't use `append()`? Couldn't find an easier way to merge two maps in go, so I wound up writing this func.

I borrowed the `append` verb from the slice equivalent (`append(slice, ...values)`) since this method seems to do more or less the same thing, but for maps instead of slices